src/device/intrinsics/math.jl

```diff
@@ -240,9 +249,12 @@ using SpecialFunctions
 # NOTE: CUDA follows fmod, which behaves differently than Base.mod for negative numbers
 #@device_override Base.mod(x::Float64, y::Float64) = ccall("extern __nv_fmod", llvmcall, Cdouble, (Cdouble, Cdouble), x, y)
 #@device_override Base.mod(x::Float32, y::Float32) = ccall("extern __nv_fmodf", llvmcall, Cfloat, (Cfloat, Cfloat), x, y)
+@device_override Base.mod(x::Float16, y::Float16) = Float16(mod(Float32(x), Float32(y)))
+
 
 @device_override Base.rem(x::Float64, y::Float64) = ccall("extern __nv_remainder", llvmcall, Cdouble, (Cdouble, Cdouble), x, y)
 @device_override Base.rem(x::Float32, y::Float32) = ccall("extern __nv_remainderf", llvmcall, Cfloat, (Cfloat, Cfloat), x, y)
+@device_override Base.rem(x::Float16, y::Float16) = Float16(rem(Float32(x), Float32(y)))
```

Comment:
Why is this required? LLVM seems to handle our `rem_float` intrinsic just fine:

```
julia> CUDA.code_llvm(rem, Tuple{Float16, Float16})
;  @ float.jl:345 within `rem'
define half @julia_rem_1728(half %0, half %1) {
top:
  %2 = frem half %0, %1
  ret half %2
}

julia> CUDA.code_ptx(rem, Tuple{Float16, Float16})
//
// Generated by LLVM NVPTX Back-End
//

.version 6.3
.target sm_75
.address_size 64

        // .globl       julia_rem_1742          // -- Begin function julia_rem_1742
.visible .func  (.param .b32 func_retval0) julia_rem_1742
(
        .param .b32 julia_rem_1742_param_0,
        .param .b32 julia_rem_1742_param_1
)
;
.weak .global .align 8 .u64 julia_rem_1742_slot = julia_rem_1742;
                                        // @julia_rem_1742
.visible .func  (.param .b32 func_retval0) julia_rem_1742(
        .param .b32 julia_rem_1742_param_0,
        .param .b32 julia_rem_1742_param_1
)
{
        .reg .b16       %h<4>;
        .reg .f32       %f<7>;

// %bb.0:                               // %top
        ld.param.b16    %h1, [julia_rem_1742_param_0];
        ld.param.b16    %h2, [julia_rem_1742_param_1];
        cvt.f32.f16     %f1, %h2;
        cvt.f32.f16     %f2, %h1;
        div.rn.f32      %f3, %f2, %f1;
        cvt.rmi.f32.f32         %f4, %f3;
        mul.f32         %f5, %f4, %f1;
        sub.f32         %f6, %f2, %f5;
        cvt.rn.f16.f32  %h3, %f6;
        st.param.b16    [func_retval0+0], %h3;
        ret;
                                        // -- End function
}
```

Reply:
Are you sure? The test case still fails for me without that definition:

```
Error in testset device/intrinsics:
Test Failed at CUDA.jl/test/device/intrinsics.jl:81
  Expression: testf((a->begin
            op.(a, T(2))
        end), T[-1])
ERROR: LoadError: Test run finished with errors
in expression starting at CUDA.jl/test/runtests.jl:490
```

`Base.mod` works though, so I've removed the `CUDA.mod` redefinition (see below).

Reply:
Ah interesting, it looks like `frem float` behaves differently than `frem half` does...